BuildingBlocks/src/BuildingBlocks.Application/MediatR/QuotaEnforcerBehavior.cs

```diff
@@ -23,21 +22,21 @@ public async Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TRe
     {
         var attributes = request.GetType().CustomAttributes;
 
-        var relevantAttribute = attributes.FirstOrDefault(attribute => attribute.AttributeType == typeof(ApplyQuotasForMetricsAttribute));
-        if (relevantAttribute != null)
+        var applyQuotasForMetricsAttribute = attributes.FirstOrDefault(attribute => attribute.AttributeType == typeof(ApplyQuotasForMetricsAttribute));
+        if (applyQuotasForMetricsAttribute != null)
         {
-            var metricKeys = relevantAttribute.ConstructorArguments.Select(it => new MetricKey(it.Value as string)).ToList();
+            var metricKeys = applyQuotasForMetricsAttribute.ConstructorArguments.Select(it => new MetricKey(it.Value as string)).ToList();
 
             var statuses = await _metricStatusesRepository.GetMetricStatuses(_userContext.GetAddress(), metricKeys);
 
-            var expiredStatuses = statuses.Where(it=> it.IsExhaustedUntil > SystemTime.UtcNow).ToList();
+            var exhaustedStatuses = statuses.Where(m => m.IsExhausted).ToList();
             
-            if (expiredStatuses.Any())
+            if (exhaustedStatuses.Any())
             {
-                var mostInTheFuture = expiredStatuses.MaxBy(it => it.IsExhaustedUntil);
-                if (mostInTheFuture != null)
+                var mostInTheFuture = exhaustedStatuses.MaxBy(it => it.IsExhaustedUntil);
```

Comment:
I just realized: there will ALWAYS be at most one MetricStatus for a given key per Identity. The key of a Metric has a unique constraint.

Reply:
I never considered there may be more than one. However, I assumed that if a command is registered with several MetricKeys (e.g. `SendMessageCommand` could be registered with `NumberOfSentMessages` and `RateOfSentMessages`) and two or more are currently echausted, it makes sense to inform the user of the one that is exhausted further in the future, i.e., the datetime after which they'll be able to use the command again. And considering this scenario, this code makes sense as far as I can see. Is my assumption wrong? I don't believe this has been cleared yet.

Reply:
Ouhh, you're right! I never thought about that scenario, tbh.

In that case, I would probably suggest to modify the QuotaExhaustedException. It should contain a list of exausted `MetricStatus`es. Caution: define a dedicated `ExhaustedMetricStatus` class for that purpose; don't use the domain class. At least for the properties. It's okay to have them for the constructor parameter.

So:

```c#
public class QuotaExhaustedException : ApplicationException
{
    public QuotaExhaustedException(MetricStatus[] exhaustedMetricStatuses)
    {
        ExhaustedMetricStatuses = exhaustedMetricStatuses.Select(...);
    }

    public ExhaustedMetricStatus[] ExhaustedMetricStatuses { get; }
}

public class ExhaustedMetricStatus 
{
    public string MetricKey { get; }
    public DateTime IsExhaustedUntil { get; }
```